File: lib/coach.ts
import type { AngleData, CoachMessage, CompensationData, ExerciseDTO, TargetAngles } from '@/types';

const SYSTEM_PROMPT = `Tu es un coach kinésithérapeute virtuel. 
Tu observes les angles articulaires d'un patient en temps réel pendant sa rééducation.
Réponds UNIQUEMENT avec un message court (max 20 mots), direct, en français.
Ton ton est encourageant mais précis. Jamais de formules de politesse.
Si tout est correct → encourage. Si déviation → corrige avec précision (quel angle, combien de degrés).
Format de réponse : texte brut uniquement, pas de markdown.`;

export type CoachProviderKey = 'gemini' | 'grok' | 'groq';

const PROVIDERS = {
  gemini: {
    model: 'gemini-2.0-flash',
    apiUrl: 'https://generativelanguage.googleapis.com/v1beta/models',
    apiKeyEnv: 'GEMINI_API_KEY',
  },
  grok: {
    model: 'grok-3-mini',
    apiUrl: 'https://api.x.ai/v1/chat/completions',
    apiKeyEnv: 'XAI_API_KEY',
  },
  groq: {
    model: 'llama-3.3-70b-versatile',
    apiUrl: 'https://api.groq.com/openai/v1/chat/completions',
    apiKeyEnv: 'GROQ_API_KEY',
  },
} as const;

export interface CoachParams {
  exercise: Pick<ExerciseDTO, 'name' | 'targetAngles'>;
  angles: AngleData;
  compensations: CompensationData;
  history: CoachMessage[];
}

function buildPrompt(params: CoachParams): string {
  return `Exercice: ${params.exercise.name}
Angles actuels: ${JSON.stringify(params.angles)}
Angles cibles: ${JSON.stringify(params.exercise.targetAngles)}
Compensations détectées: ${JSON.stringify(params.compensations)}
Derniers messages envoyés: ${params.history
    .slice(-3)
    .map((m) => m.text)
    .join(' | ')}`;
}

async function callGemini(
  config: { model: string; apiUrl: string; apiKey: string },
  userContent: string,
): Promise<string> {
  const res = await fetch(`${config.apiUrl}/${config.model}:generateContent?key=${config.apiKey}`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      system_instruction: { parts: [{ text: SYSTEM_PROMPT }] },
      contents: [{ role: 'user', parts: [{ text: userContent }] }],
      generationConfig: { maxOutputTokens: 150 },
    }),
  });
  if (!res.ok) {
    const err = await res.text();
    throw new Error(`Gemini API error: ${err}`);
  }
  const data = (await res.json()) as {
    candidates?: { content?: { parts?: { text?: string }[] } }[];
  };
  const text = data.candidates?.[0]?.content?.parts?.[0]?.text;
  if (!text) throw new Error('Réponse Gemini vide');
  return text.trim();
}

async function callOpenAICompatible(
  config: { model: string; apiUrl: string; apiKey: string },
  userContent: string,
): Promise<string> {
  const res = await fetch(config.apiUrl, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${config.apiKey}`,
    },
    body: JSON.stringify({
      model: config.model,
      max_tokens: 150,
      messages: [
        { role: 'system', content: SYSTEM_PROMPT },
        { role: 'user', content: userContent },
      ],
    }),
  });
  if (!res.ok) {
    const err = await res.text();
    throw new Error(`API error: ${err}`);
  }
  const data = (await res.json()) as {
    choices?: { message?: { content?: string } }[];
  };
  const text = data.choices?.[0]?.message?.content;
  if (!text) throw new Error('Réponse API vide');
  return text.trim();
}

const ADAPTERS = {
  gemini: callGemini,
  grok: callOpenAICompatible,
  groq: callOpenAICompatible,
} as const;

export function inferMessageType(
  angles: AngleData,
  targets: TargetAngles,
  compensations: CompensationData,
): 'info' | 'warn' | 'success' {
  if (compensations.lumbar > 0 || compensations.shoulder > 0) return 'warn';
  const keys = Object.keys(targets) as (keyof AngleData)[];
  const outOfTarget = keys.some((key) => {
    const t = targets[key as string];
    if (t === undefined) return false;
    return Math.abs((angles[key] ?? 0) - t) > 10;
  });
  if (outOfTarget) return 'warn';
  return 'success';
}

export async function getCoachFeedback(
  params: CoachParams,
  provider?: CoachProviderKey,
): Promise<{ message: string; type: 'info' | 'warn' | 'success' }> {
  const selected = provider ?? (process.env.COACH_PROVIDER as CoachProviderKey) ?? 'gemini';
  const config = PROVIDERS[selected];
  if (!config) throw new Error(`Provider inconnu : ${selected}`);

  const apiKey = process.env[config.apiKeyEnv];
  if (!apiKey) throw new Error(`Clé API manquante : ${config.apiKeyEnv}`);

  const message = await ADAPTERS[selected]({ ...config, apiKey }, buildPrompt(params));
  const type = inferMessageType(params.angles, params.exercise.targetAngles, params.compensations);
  return { message, type };
}

File: core/coach.py
import os
import time

from data.exercises import Exercise
from data.session_state import AngleData, CoachMessage

SYSTEM_PROMPT = """Tu es un coach kinésithérapeute virtuel.
Réponds UNIQUEMENT avec un message court (max 20 mots), direct, en français.
Ton ton est encourageant mais précis. Jamais de formules de politesse.
Si tout est correct → encourage. Si déviation → corrige avec précision (quel angle, combien de degrés).
Format : texte brut uniquement, pas de markdown."""


class CoachAI:
    def __init__(self):
        groq_key = os.getenv("GROQ_API_KEY", "").strip()
        gemini_key = os.getenv("GEMINI_API_KEY", "").strip()

        self.groq_client = None
        self.gemini_model = None

        if groq_key:
            try:
                from groq import Groq

                self.groq_client = Groq(api_key=groq_key)
            except Exception:
                self.groq_client = None

        if gemini_key:
            try:
                import google.generativeai as genai

                genai.configure(api_key=gemini_key)
                self.gemini_model = genai.GenerativeModel("gemini-1.5-flash")
            except Exception:
                self.gemini_model = None

        self._last_call_time = 0.0
        self._use_groq = True
        self.MIN_INTERVAL = 8.0

    def should_call(self) -> bool:
        return (time.time() - self._last_call_time) >= self.MIN_INTERVAL

    def get_message(
        self,
        angles: AngleData,
        exercise: Exercise,
        compensations: dict,
        history: list[CoachMessage],
    ) -> CoachMessage | None:
        if not self.should_call():
            return None

        prompt = f"""Exercice: {exercise.name}
Angles actuels: épaule G={angles.shoulder_left:.0f}°, épaule D={angles.shoulder_right:.0f}°, coude G={angles.elbow_left:.0f}°, colonne={angles.spine:.0f}°
Angles cibles: {exercise.target_angles}
Compensations détectées: {compensations}
Derniers messages: {[m.text for m in history[-2:]]}"""

        msg_type = (
            "warn"
            if (compensations.get("lumbar", 0) > 0 or compensations.get("shoulder", 0) > 0)
            else "info"
        )

        if not self.groq_client and not self.gemini_model:
            self._last_call_time = time.time()
            return self._static_feedback(angles, exercise, compensations)

        try:
            text = self._call_groq(prompt) if self._use_groq else self._call_gemini(prompt)
            self._use_groq = not self._use_groq
            self._last_call_time = time.time()
            return CoachMessage(text=text, msg_type=msg_type)
        except Exception:
            try:
                text = self._call_gemini(prompt) if self._use_groq else self._call_groq(prompt)
                self._last_call_time = time.time()
                return CoachMessage(text=text, msg_type=msg_type)
            except Exception:
                self._last_call_time = time.time()
                return self._static_feedback(angles, exercise, compensations)

    def _call_groq(self, prompt: str) -> str:
        if not self.groq_client:
            raise RuntimeError("Groq client unavailable")
        response = self.groq_client.chat.completions.create(
            model="llama-3.1-8b-instant",
            max_tokens=60,
            messages=[
                {"role": "system", "content": SYSTEM_PROMPT},
                {"role": "user", "content": prompt},
            ],
        )
        return response.choices[0].message.content.strip()

    def _call_gemini(self, prompt: str) -> str:
        if not self.gemini_model:
            raise RuntimeError("Gemini model unavailable")
        response = self.gemini_model.generate_content(
            f"{SYSTEM_PROMPT}\n\n{prompt}",
            generation_config={"max_output_tokens": 60},
        )
        return response.text.strip()

    def _static_feedback(
        self, angles: AngleData, exercise: Exercise, compensations: dict
    ) -> CoachMessage:
        if compensations.get("lumbar", 0) > 0:
            return CoachMessage(
                text="Compensation lombaire — contractez les abdominaux.", msg_type="warn"
            )
        target = exercise.target_angles
        diff = abs(angles.shoulder_left - target.get("shoulder_left", 90))
        if diff > 15:
            return CoachMessage(
                text=f"Épaule gauche : ajustez de {diff:.0f}°.", msg_type="warn"
            )
        return CoachMessage(text="Bonne exécution. Continuez ce rythme.", msg_type="success")
